Games resource

gianluca-nitti opened the pull request:
## Contents
The code on this branch adds support for management of the savegames stored on the server from the web client/API, as described in the 16 July (week 7) [devlog update](https://gianluca-nitti.github.io/GSoC-2017-devlog/).

## Before merging
I'm marking this as WIP since there are a couple of parts that I think need to be improved before this branch is merged:
- ~~The `getLoggingPhase()` method of the `StateLoading` and `StateInGame` engine states is used to determine the name of the game which is running or loading, which, as pointed out in the comments in the code, works but in my opinion is not a very good idea, because for example the `getLoggingPhase()` might be changed in the future to provide additional information. So I'm considering opening a pull request in [MovingBlocks/Terasology](https://github.com/movingblocks/terasology) which adds a more suitable method to get this information (e.g. `getGameTitle()` or something like that).~~ Fixed in 	1a087c4 by using `currentState.getContext().get(Game.class).getName();`

- ~~As pointed out in the blog update `AvailableModulesResource` is an hack to circumvent the fact that CORS is not enabled on the [Meta Server](https://github.com/MovingBlocks/meta-server),thus the list of available modules can't be obtained directly from a web browser, so the game server is being used as a proxy. As described in the blog post we could either enable CORS on the meta server or improve `AvailableModulesResource` to provide different or additional information, for example only listing the modules installed on the game server.~~ `AvailableModulesResource` now provides information about **installed** modules and world generators.
- ~~When creating new games, any module can be specified from the web frontend, but if a module is not installed on the game server the dependency resolution fails. This is also the reason I still haven't implemented the feature to modify an existing savegame's module list (actually existing savegames can be deleted, backed up or renamed). Two possible solutions could be:~~
    - ~~Download the missing modules on-the-fly when the game is created.~~
Picked this solution in commit 2e15c95:
    - Show only the installed modules on the game creation screen, and add a separate resource class (and a new tab in the frontend application) to install new modules in the server.

Also, as pointed out at the end of the blog post, the engine can be started in this new state (`StateEngineIdle`) suggested by @msteiger by specifying the `-dontStartDefault` command line switch, but at the moment there is no way to actually avoid the creation of a default game because you can't login to the web interface without a valid client identity certificate, which is generated when connecting to a running game. The planned solution to this problem is to set up a sort of admin password that can be used to login to the web interface to only administer savegames without a client identity (but obviousluy as always any idea is welcome); let me know if I should add it on this branch or wait for this pull request to be reviewed and add it later.

## How to test
It's possible to access the new resources through the "Games and Modules" tab in the web client, you can either use the version hosted on GitHub Pages [here](https://gianluca-nitti.github.io/FacadeServer-frontend/) or clone the [repository](https://github.com/gianluca-nitti/facadeserver-frontend) locally and following the build instructions in the README (basically, you need to have `npm` installed, install the dependencies including the development ones and run `npm run web`). I'm aware that there is still some small usability problem in the frontend application but it shouldn't be a problem when using it for testing the server. ~~For example, if you connect to an not existing server, no warning is shown and various buttons become unresponsive (as pointed out by @skaldarnar last week). This week I mainly focused on the server and I'm going to fix this soon, in the meanwhile please press the "Connect" button on the "Server Address" dialog only when the local server is running and has completed initialization.~~ Fixed.

gianluca-nitti committed:
Initial implementation of game metadata resource

gianluca-nitti committed:
Add engine status resource

gianluca-nitti committed:
Allow authorized clients to change engine state

gianluca-nitti committed:
Handle exceptions thrown when reading/writing resources

gianluca-nitti committed:
Allow access to games resource when a game is running too

gianluca-nitti committed:
Notify clients of ResourceManager reinitialize for engine state changes

gianluca-nitti committed:
Fix authentication after engine state change

gianluca-nitti committed:
Fix start and stop of savegames

gianluca-nitti committed:
First implementation of savegame manipulation classes

gianluca-nitti committed:
Improve server admin permission management

gianluca-nitti committed:
Fix savegame backup and savegame list change notifications

gianluca-nitti committed:
Disallow backup or deletion of the running game

gianluca-nitti committed:
Add support for new game generation

gianluca-nitti committed:
Add game renaming feature

gianluca-nitti committed:
Add unit tests for game management classes

gianluca-nitti committed:
Refactor EngineRunner to a singleton to improve unit tests

gianluca-nitti committed:
Fixed various Checkstyle warnings

gianluca-nitti committed:
Provide information about installed modules and world generators

gianluca-nitti committed:
Fix code to retrieve the running game's name

gianluca-nitti committed:
Refactor: use dependency injection in all resource classes

gianluca-nitti committed:
Add unit test for AvailableModulesResource class

gianluca-nitti committed:
Small fix: remove duplicated resources in ResourceManager

msteiger commented on src/main/java/org/terasology/web/resources/games/AbstractAction.java:
```diff
@@ -0,0 +1,41 @@
+import org.terasology.engine.paths.PathManager;
+import org.terasology.web.resources.ResourceAccessException;
+
+public abstract class AbstractAction {
```
I always try to make an abstract class implement an interface, to allow implementors to provide alternative content. Since this class just provides trivial code, it could be moved into the actual implementation and converted directly into an interface. But this would require removing the `setPathManager` method for the tests. Hmm .. tricky.

gianluca-nitti replied:
Okay, refactored the class structure a bit and converted this to an interface in latest commit (373bbfd).

msteiger commented on src/main/java/org/terasology/web/resources/games/BackupGameAction.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Override
+    public void perform(String gameName) throws ResourceAccessException {
+        String backupName = gameName + "_backup_" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy_MM_dd_HH_mm_ss"));
```
The formatter could be a constant

msteiger commented on src/main/java/org/terasology/web/resources/games/NewGameAction.java:
```diff
@@ -0,0 +1,101 @@
+        return newGameManifest;
+    }
+
+    void setFields(String fGameName, String fSeed, List<Name> fModules, SimpleUri fWorldGenerator) {
```
Looks like something that is useful for testing. Would it make sense to convert this into a package private constructor?

gianluca-nitti replied:
Converted to a constructor as part of commit 373bbfd.

msteiger commented on src/main/java/org/terasology/web/resources/games/NewGameAction.java:
```diff
@@ -0,0 +1,101 @@
+        }
+
+        WorldInfo worldInfo = new WorldInfo(TerasologyConstants.MAIN_WORLD, newGameManifest.getSeed(),
+                (long) (WorldTime.DAY_LENGTH * 0.025f), worldGenerator);
```
This looks like copy&paste from some other place. I suspect that it wasn't possible to reuse the code. This world time setting is something for a different PR, I guess.

gianluca-nitti replied:
Yes, these 2 lines are taken from [StateHeadlessSetup](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/engine/subsystem/headless/mode/StateHeadlessSetup.java#L141); the time offset is set to that value to avoid the player spawning in a completely dark environment. In my understanding though it's not possible to reuse the whole [createGameManifest](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/engine/subsystem/headless/mode/StateHeadlessSetup.java#L109) method since it just uses the default settings for title, seed and world generator while here we are using user-supplied values.

As you suggest, in another branch I could implement a Resource to allow to view and/or manipulate the world time of an existing savegame or maybe the currently running game. In my understanding I still need to set the world time to some value when initializing the game manifest, I could set it to zero if you prefer?

msteiger replied:
No, it's fine. Sorry for distracting you. This world-time setting is set at various places and I'd like to centralize at some time, but that's a different story.

msteiger commented on src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
What is this?

gianluca-nitti replied:
It enables an incubating feature of mockito to mock final classes, which is done in some of the tests: https://github.com/mockito/mockito/wiki/What's-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods

msteiger replied:
Ah, that's good to know. Thanks!

msteiger reviewed (approved):
Very nice, good work @gianluca-nitti

gianluca-nitti committed:
Refactor GamesResource action classes.

This improves readability, and the AbstractAction abstract class has been converted to an interface (new name is GameAction) as suggested by @msteiger in review of PR #4.

msteiger commented:
Thank you very much!
